WooCommerce/Classes/ViewRelated/Notifications/ReviewsViewModel.swift

```diff
@@ -0,0 +1,128 @@
+
+    var dataSource: UITableViewDataSource {
+        return data
+    }
+
+    var delegate: UITableViewDelegate {
+        return data
+    }
+
+    init(data: ReviewsDataSource) {
+        self.data = data
+    }
+
+    func displayPlaceholderReviews(tableView: UITableView) {
+        let options = GhostOptions(reuseIdentifier: ProductReviewTableViewCell.reuseIdentifier, rowsPerSection: Settings.placeholderRowsPerSection)
+        tableView.displayGhostContent(options: options)
+
+        data.stopForwardingEvents()
+    }
+
+    /// Removes Placeholder Notes (and restores the ResultsController <> UITableView link).
+    ///
+    func removePlaceholderReviews(tableView: UITableView) {
+        tableView.removeGhostContent()
+        data.startForwardingEvents(to: tableView)
+    }
+
+    func configureResultsController(tableView: UITableView) {
+        data.startForwardingEvents(to: tableView)
+        try? data.observeReviews()
+    }
+
+    /// Setup: TableViewCells
+    ///
+    func configureTableViewCells(tableView: UITableView) {
+        let cells = [ProductReviewTableViewCell.self]
+
+        for cell in cells {
+            tableView.register(cell.loadNib(), forCellReuseIdentifier: cell.reuseIdentifier)
+        }
+    }
+}
+
+
+// MARK: - Fetching data
+extension ReviewsViewModel {
+    /// Prepares data necessary to render the reviews tab.
+    ///
+    func synchronizeReviews(onCompletion: (() -> Void)? = nil) {
+        let group = DispatchGroup()
+
+        group.enter()
+        synchronizeAllReviews {
+            group.leave()
+        }
+
+        group.enter()
+        synchronizeProductsReviewed {
+            group.leave()
+        }
+
+        group.notify(queue: .main) {
+            if let completionBlock = onCompletion {
+                completionBlock()
+            }
+        }
+    }
+
+    /// Synchronizes the Reviews associated to the current store.
+    ///
+    private func synchronizeAllReviews(onCompletion: (() -> Void)? = nil) {
+        guard let siteID = ServiceLocator.stores.sessionManager.defaultStoreID else {
+            return
+        }
+
+        let action = ProductReviewAction.synchronizeProductReviews(siteID: siteID, pageNumber: 1, pageSize: 25) { error in
+            if let error = error {
+                DDLogError("⛔️ Error synchronizing reviews: \(error)")
+            } else {
+                //TODO. What event must be sent here?
+                //ServiceLocator.analytics.track(.notificationListLoaded)
+            }
+
+            onCompletion?()
+        }
+
+        ServiceLocator.stores.dispatch(action)
+    }
+
+    private func synchronizeProductsReviewed(onCompletion: @escaping () -> Void) {
+        let reviewsProductIDs = data.reviewsProductsIDs
+
+        guard let siteID = ServiceLocator.stores.sessionManager.defaultStoreID else {
+            return
+        }
+
+        let action = ProductAction.retrieveProducts(siteID: siteID, productIDs: reviewsProductIDs) { error in
```

Comment:
What would be the advantage of that exactly? (I'm trying to fully understand the idea)

Reply:
So...is anything listening for the updates that happen after the products are retrieved? I didn't see the UI refresh itself when an unknown (not previously downloaded) product came in as a review:

<img src="https://user-images.githubusercontent.com/1062444/64290132-23a3ee00-cf2b-11e9-88c3-ee7f33806d1c.png" width="350" />

I did eventually use pull-to-refresh and saw this:
<img src="https://user-images.githubusercontent.com/1062444/64290163-31597380-cf2b-11e9-9674-6b32afd1eddd.png" width="350" />

Reply:
> What would be the advantage of that exactly? (I'm trying to fully understand the idea)

if a `ProductReview` model has a relationship to `Product`, then we can just get a review's product by `productReview.product` instead of manually mapping `productReview.productID` to a `Product`